src/client/Fig.Client/ConfigurationProvider/FigConfigurationBuilder.cs

```
@@ -0,0 +1,84 @@
+
+    public IDictionary<string, object> Properties => _configurationBuilder.Properties;
+
+    public IList<IConfigurationSource> Sources => _configurationBuilder.Sources;
+
+    public FigConfigurationBuilder(IConfigurationBuilder configurationBuilder, FigOptions figOptions, Type settingsType)
+    {
+        _configurationBuilder = configurationBuilder ?? throw new ArgumentNullException(nameof(configurationBuilder));
+        _figOptions = figOptions ?? throw new ArgumentNullException(nameof(figOptions));
+        _settingsType = settingsType;
+        ValidateClientName();
+    }
+
+    public IConfigurationBuilder Add(IConfigurationSource source)
+    {
+        return _configurationBuilder.Add(source);
+    }
+
+    public IConfigurationRoot Build()
+    {
+        var source = new FigConfigurationSource
+        {
+            LoggerFactory = _figOptions.LoggerFactory,
+            ApiUri = ReadFigApiFromEnvironmentVariable(),
+            PollIntervalMs = _figOptions.PollIntervalMs,
+            LiveReload = _figOptions.LiveReload,
+            Instance = ReadInstanceFromEnvironmentVariable(_figOptions.ClientName),
+            ClientName = _figOptions.ClientName,
+            VersionOverride = _figOptions.VersionOverride,
+            AllowOfflineSettings = _figOptions.AllowOfflineSettings,
+            SettingsType = _settingsType,
+            SupportsRestart = _figOptions.SupportsRestart,
+            HttpClient = _figOptions.HttpClient,
+            ClientSecretOverride = _figOptions.ClientSecretOverride
+        };
+
+        Logging.Logger.LoggerFactory = source.LoggerFactory ?? new NullLoggerFactory();
+        var logger = Logging.Logger.CreateLogger<FigConfigurationBuilder>();
+
+        if (source.HttpClient is null && (string.IsNullOrWhiteSpace(source.ApiUri) || !Uri.TryCreate(source.ApiUri, UriKind.Absolute, out _)))
+        {
+            logger.LogWarning("Empty or invalid Fig API URI. Fig configuration provider will be disabled.");

```

Better to throw an exception here? Can't quite follow it through, but assuming you're intending to AddFig it seems pretty critical if it won't work.

---

So [here ](https://github.com/mzbrau/fig/pull/3/files#diff-d5f53a191fc88b0585032b4a66fd5ac4c8e6d59b4342c523c0b7db112b7a60cdR23)I have a recommendation that Fig be registered as the last configuration provider. The thinking here is that you don't want another configuration provider to overwrite a value that looks like it is being set by Fig. That will be very confusing for the person configuring. However you might want to run the application without Fig being installed so you probably want to have the options of using appSettings.json files. My thinking is that the FIG_API_URI environment variable is essentially a switch. If you don't have that present on your machine, then Fig will be disabled and you can use any other configuration provider. If you do have it present, then Fig will initialize and have precedence over configuration.